Latest NW Monitoring Enhancements

jlquaccia opened the pull request:
- user device detection (mobile, tablet, desktop)
- capturing network type (only for supported browsers, will not be captured for browsers that do not support this feature)
- wrapped entire onPage script in an IIFE
- keeping track of the sequence of order that bidder calls are made each auction
- keeping track of the number of ad units that are present during each auction
- keeping track of the request size (collecting the char count), for each bidder (per auction)
- keeping track of the bidder req call type (post, get, etc.) (per auction)
- bidders to keep track of during auctions are now dynamically set based on who is participating in an auction
- ensuring that the cf worker only gets called once per bidder but also keeps track of all network calls for each bid a bidder may have
- keeping track of perceived latency for each bidder bid request
- keeping track of bidder bid responses (per auction), whether or not they were used in the auction or not
- keeping track of the number of bid calls a bidder makes (per auction)
- dropped the `key` field (ex: `key: "pm"`) from each object within the bidders array
- Slimmed down the json payload size sent to the cf worker by truncating the perf resource url's to not include any query params

jlquaccia commented:
Hey @pm-harshad-mane, after discussing with @pm-nitin-shirsat, we both just had a couple of questions around the logic for this PR.  Included questions in the comments below:

jlquaccia commented on OnPageCode_timeout.js:
```diff
@@ -1,288 +1,457 @@
-      db: perfResource?.name?.length ? 0 : 1
-    };
+      };
+      currentBidCall.requestUrlPayloadLength = bidderRequest?.nwMonitor?.requestUrlPayloadLength;
```
Line 309 is specific to PubMatic and captures the payload length (which includes the pubmatic url and payload).  

Line 310 is taken from pbjs and only captures the length of the payload (not additionally including the length of the url too, just the payload).  This is captured for all bid requests from pbjs.

Question is, how should we capture this payload length data?  should we capture this within 2 separate fields?  or under one field, maybe something like the following?

```
currentBidCall.reqPayloadLength =
    bidderRequest?.nwMonitor?.requestUrlPayloadLength
        ? bidderRequest?.nwMonitor?.requestUrlPayloadLength
        : sspConf.bids[processedBidsCount - 1].reqPayloadCharCount;
```

jlquaccia commented on OnPageCode_timeout.js:
```diff
@@ -1,288 +1,457 @@
-      timeEndKey: "responseEnd",
-      timeStartKey: "responseStart"
+
+    'bidders': [],
```
In terms of dynamically adding bidders in this script to track nw data for them per auction, @pm-nitin-shirsat brought up a good point.. as of right now all bidders in an auction will be included in this script, but some auctions may have a large amount of bidders..

Should we whitelist bidders that we would want to dynamically include in this script?  or would we want to just include all participating bidders each time in an auction?

pm-nitin-shirsat replied:
@jlquaccia - As I discussed this point with Deepak, he suggested, we can keep dynamic bidders as we are using one or two publishers for our testing and which can have max 10 bidders each. 
But I think, if we need analysis further we may need more publishers which will increase the space in ELK. so it's better to whitelist the bidders. 
your thoughts @pm-harshad-mane ??

jlquaccia commented:
Updated this PR to include the following changes:

- Refactored payload schema that gets passed to ELK (also now including auctionId)
- Passing payloads to ELK once per bid request (previously was once per bidder)
- Created a whitelist for which bidders are allowed to dynamically participate in network monitoring during an auction
- Fixed bug where nw auction data reset issue when multiple auctions occur close to one another
- Fixed processed bid count bug

pm-nitin-shirsat reviewed (approved):
Hi @jlquaccia - I have reviewed and verified the changes on different pages and its works fine in all cases. Good to go from my side.
